craft_parts/parts.py

```diff
@@ -64,36 +81,39 @@ class PartSpec(BaseModel):
     override_stage: Optional[str] = None
     override_prime: Optional[str] = None
     permissions: List[Permissions] = []
-
-    class Config:
-        """Pydantic model configuration."""
-
-        validate_assignment = True
-        extra = "forbid"
-        allow_mutation = False
-        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+    model_config = ConfigDict(
+        validate_assignment=True,
+        extra="forbid",
+        frozen=True,
+        strict=True,
+        alias_generator=lambda s: s.replace("_", "-"),
+    )
 
     # pylint: disable=no-self-argument
-    @validator("stage_files", "prime_files", each_item=True)
-    def validate_relative_path_list(cls, item: str) -> str:
+    @field_validator("stage_files", "prime_files")
+    @classmethod
+    def validate_relative_path_list(cls, itemlist: List[str]) -> List[str]:
         """Verify list is not empty and does not contain any absolute paths."""
-        if not item:
-            raise ValueError("path cannot be empty")
-        if item.startswith("/"):
-            raise ValueError(
-                f"{item!r} must be a relative path (cannot start with '/')"
-            )
-        return item
-
-    @validator("overlay_packages", "overlay_files", "overlay_script")
+        for item in itemlist:
+            if not item:
+                raise ValueError("path cannot be empty")
+            if item.startswith("/"):
+                raise ValueError(
+                    f"{item!r} must be a relative path (cannot start with '/')"
+                )
+        return itemlist
+
+    @field_validator("overlay_packages", "overlay_files", "overlay_script")
+    @classmethod
     def validate_overlay_feature(cls, item: Any) -> Any:  # noqa: ANN401
         """Check if overlay attributes specified when feature is disabled."""
         if not Features().enable_overlay:
             raise ValueError("overlays not supported")
         return item
 
-    @root_validator(pre=True)
-    def validate_root(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+    @model_validator(mode="after")
+    @classmethod
+    def validate_root(cls, values: Any) -> Any:  # noqa: ANN401
```

Comment:
Will adjust that. Thanks for pointing it out!

Reply:
shouldn't this be `mode="before"`, to keep the semantics of `pre=True`?

Reply:
I tried using non-class methods in `mode=after` validators as pointed out in the referenced doc, however it seems to be requiring a considerable amount of change in the asserts in our tests for it to work without error. Using class methods works just fine, although the doc implies it shouldn't, as we're using `cast` to get the properties from `values` parameter instead of getting them through `self`.